SU2_CFD/src/output/CAdjElasticityOutput.cpp

```diff
@@ -0,0 +1,177 @@
+  }
+
+  if (nRequestedVolumeFields == 0){
+    RequestedVolumeFields.push_back("COORDINATES");
+    RequestedVolumeFields.push_back("SOLUTION");
+    nRequestedVolumeFields = RequestedVolumeFields.size();
+  }
+
+  stringstream ss;
+  ss << "Zone " << config->GetiZone() << " (Adj. Comp. Fluid)";
+  MultiZoneHeaderString = ss.str();
+
+  /*--- Set the volume filename --- */
+  
+  VolumeFilename = config->GetAdj_FileName();
+  
+  /*--- Set the surface filename --- */
+  
+  SurfaceFilename = config->GetSurfAdjCoeff_FileName();
+  
+  /*--- Set the restart filename --- */
+  
+  RestartFilename = config->GetRestart_FileName();
+
+  /*--- Set the default convergence field --- */
+
+  if (Conv_Field.size() == 0 ) Conv_Field = "ADJOINT_DISP_X";
+  
+}
+
+CAdjElasticityOutput::~CAdjElasticityOutput(void) {
+
+  if (rank == MASTER_NODE){
+    HistFile.close();
+  }
+
+
+}
+
+void CAdjElasticityOutput::SetHistoryOutputFields(CConfig *config){
+  
+  // Residuals
+  AddHistoryOutput("ADJOINT_DISP_X", "Res[Ux_adj]", FORMAT_FIXED,   "RESIDUALS");
+  AddHistoryOutput("ADJOINT_DISP_Y", "Res[Uy_adj]", FORMAT_FIXED,   "RESIDUALS");
+  AddHistoryOutput("ADJOINT_DISP_Z", "Res[Uz_adj]", FORMAT_FIXED,   "RESIDUALS");
+  
+  //Sensitivities
+  AddHistoryOutput("SENS_E", "Sens[E]",  FORMAT_FIXED, "SENSITIVITY");
+  AddHistoryOutput("SENS_NU","Sens[Nu]", FORMAT_FIXED, "SENSITIVITY");
+
+  
+}
+
+inline void CAdjElasticityOutput::LoadHistoryData(CConfig *config, CGeometry *geometry, CSolver **solver) {
+  
+  SetHistoryOutputValue("ADJOINT_DISP_X", log10(solver[ADJFEA_SOL]->GetRes_RMS(0)));
+  SetHistoryOutputValue("ADJOINT_DISP_Y", log10(solver[ADJFEA_SOL]->GetRes_RMS(1)));
+  if (nVar_FEM == 3){
+    SetHistoryOutputValue("ADJOINT_DISP_Z", log10(solver[ADJFEA_SOL]->GetRes_RMS(2)));    
+  }
+  su2double Total_SensE = 0.0; su2double Total_SensNu = 0.0;  
+  if (config->GetnElasticityMod() == 1){
+    Total_SensE = solver[ADJFEA_SOL]->GetGlobal_Sens_E(0);
+    Total_SensNu = solver[ADJFEA_SOL]->GetGlobal_Sens_Nu(0);
+  }
+  else{
+    // TODO: Update this and change tests
+    for (unsigned short iVar = 0; iVar < config->GetnElasticityMod(); iVar++){
+      Total_SensE += solver[ADJFEA_SOL]->GetGlobal_Sens_E(0)
+                    *solver[ADJFEA_SOL]->GetGlobal_Sens_E(0);
+      Total_SensNu += solver[ADJFEA_SOL]->GetGlobal_Sens_Nu(0)
+                     *solver[ADJFEA_SOL]->GetGlobal_Sens_Nu(0);
+    }
+  Total_SensE = sqrt(Total_SensE);
+  Total_SensNu = sqrt(Total_SensNu);
+  }
+  SetHistoryOutputValue("SENS_E", Total_SensE);
+  SetHistoryOutputValue("SENS_NU", Total_SensNu);
+  
+}
+
+void CAdjElasticityOutput::LoadVolumeData(CConfig *config, CGeometry *geometry, CSolver **solver, unsigned long iPoint){
+
+  CVariable* Node_Struc = solver[FEA_SOL]->node[iPoint];
+  CPoint*    Node_Geo  = geometry->node[iPoint];
+
+  SetVolumeOutputValue("COORD-X", iPoint,  Node_Geo->GetCoord(0));
+  SetVolumeOutputValue("COORD-Y", iPoint,  Node_Geo->GetCoord(1));
```

Comment:
You are hashing into the map a bunch of times per node but the access sequence is always the same, you could cache the offset (column?) during preprocessing and not go through the map here.

Reply:
Thanks for mentioning this! That was indeed not very efficient. I added caching of the offset. That speeds up the whole process a lot. I implemented it in a way so that the order the SetVolumeOutputValue calls can be still different from the order of the AddVolumeOutput calls.

Reply:
Looks good @talbring, should do the trick provided the function is not called twice for the same variable.
Will the list of output variables be exposed to the config? (I remember from yours and @rsanfer 's presentation that the screen output will be) Is that why the variables are being referenced via strings? Have you considered a pair of enum/map like we have for other options? That way you could create a vector of enum instead of a vector of strings and use that as a "map" which would be efficient without needing to rely on access order, etc.

Reply:
What do you mean exactly by "exposing it to the config" ? The rational behind having it as strings is that it is independent of the config structure. So just adding output doesn't require to recompile the whole code. 

I added a further check to make sure that the SetVolumeOutputValue() was not called twice for the same field. The good thing is that by adding these checks I already found some bugs :)

Reply:
I mean having a config option that allows the user to pick which fields they want in the output file.

Reply:
Oh thats what you mean! You can specify it in the config file. That was also an important point of the implementation. 

There are the options
```
VOLUME_OUTPUT
SCREEN_OUTPUT
HISTORY_OUTPUT
```
to specify the fields you want to have. The config structure just has a string list for these options and no enum/map combination.

A more detailed description follows soon.

Reply:
Ok I'll hold further comments until then, but if the concern with the enum/map pair is compilation dependencies those can be defined at the level of the COutput children classes.